pkg/session/session.go

```diff
@@ -2379,6 +2379,39 @@ func (rs *execStmtResult) Close() error {
 	return err2
 }
 
+func (rs *execStmtResult) Detach() (sqlexec.RecordSet, error) {
+	if !rs.sql.IsReadOnly(rs.se.GetSessionVars()) {
+		return nil, errors.New("cannot detach record set for non-read-only statement")
+	}
+	if !plannercore.IsAutoCommitTxn(rs.se.GetSessionVars()) {
+		return nil, errors.New("cannot detach record set in non-auto-commit transaction")
+	}
+
+	drs, ok := rs.RecordSet.(sqlexec.DetachableRecordSet)
+	if !ok {
+		return nil, errors.New("record set cannot be detached")
+	}
+	detachedRS, err := drs.Detach()
+	if err != nil {
+		return nil, err
+	}
+
+	// FIXME: if the following code meets error, the `detachedRS` will be leaked.
+	// The error may occur during the begin/commit of the newly created transaction on the given `startTS``.
+
+	// FIXME: block the min-start-ts. Now the `min-start-ts` will advance and exceed the `startTS` used by
+	// this detached record set, and may cause an error if `GC` runs.
+
+	// Now, a transaction is not needed for the detached record set, so we commit the transaction and cleanup
+	// the session state.
+	err = rs.se.CommitTxn(context.Background())
```

Comment:
And what is the expected operation for an old RecordSet if a developer called `Detach`? Does he still need to call the `Close` method of the old one? If it is yes, I think we do not need to call `CommitTxn` because the old one does it.

Reply:
> Does he still need to call the `Close` method of the old one?

No. If the `Detach` return successfully, the old `RecordSet` shouldn't be used anymore (including `Close()`).

I think it's not easy to write a `Close` suitable for both detached and not-detached record set. Therefore, we should clean up TXN states here.

Reply:
I'll use `finishStmt` directly.